src/go/k8s/pkg/resources/secret.go

```diff
@@ -0,0 +1,247 @@
+	return fmt.Sprintf(`#!/usr/bin/env bash
+set -e
+
+until NODE_ID=$(%s | grep -o '\"node_id\":[^,}]*' | grep -o '[^: ]*$'); do
+	sleep 0.5
+done
+echo "Clearing maintenance mode on node ${NODE_ID}"
+until [ "${status:-}" = "200" ] || [ "${status:-}" = "400" ]; do
+	status=$(%s)
+	sleep 0.5
+done`, curlNodeIDCommand, curlCommand)
+}
+
+// getPrestopScript creates a script that drains the node before shutting down.
+func (r *SecretResource) getPreStopScript() string {
+	// TODO replace scripts with proper RPK calls
+	curlNodeIDCommand := r.composeCURLGetNodeIDCommand("--silent --fail")
+	curlMaintenanceCommand := r.composeCURLMaintenanceCommand(`-X PUT --silent -o /dev/null -w "%{http_code}"`, nil)
+	genericMaintenancePath := "/v1/maintenance"
+	curlGetMaintenanceCommand := r.composeCURLMaintenanceCommand(`--silent`, &genericMaintenancePath)
+	// TODO(joe): The command builder shouldn't add auth/tls flags, they should be added to an environment variable
+	// in the StatefulSet otherwise a race condition could occur where the mounted script could be updated before
+	// the new setting has been configured on the pod's redpanda node.
+	return fmt.Sprintf(`#!/usr/bin/env bash
+set -e
+
+until NODE_ID=$(%s | grep -o '\"node_id\":[^,}]*' | grep -o '[^: ]*$'); do
+	sleep 0.5
+done
+echo "Setting maintenance mode on node ${NODE_ID}"
+until [ "${status:-}" = "200" ]; do
+	status=$(%s)
+	sleep 0.5
+done
+until [ "${finished:-}" = "true" ] || [ "${draining:-}" = "false" ]; do
+	res=$(%s)
+	finished=$(echo $res | grep -o '\"finished\":[^,}]*' | grep -o '[^: ]*$')
+	draining=$(echo $res | grep -o '\"draining\":[^,}]*' | grep -o '[^: ]*$')
+	sleep 0.5
+done`, curlNodeIDCommand, curlMaintenanceCommand, curlGetMaintenanceCommand)
+}
+
+//nolint:goconst // no need
+func (r *SecretResource) composeCURLGetNodeIDCommand(
+	options string,
+) string {
+	adminAPI := r.pandaCluster.AdminAPIInternal()
+
+	cmd := fmt.Sprintf(`curl %s `, options)
+
+	tlsConfig := adminAPI.GetTLS()
+	proto := "http"
+	if tlsConfig != nil && tlsConfig.Enabled {
+		proto = "https"
+		if tlsConfig.RequireClientAuth {
+			cmd += "--cacert /etc/tls/certs/admin/ca/ca.crt --cert /etc/tls/certs/admin/tls.crt --key /etc/tls/certs/admin/tls.key "
+		} else {
+			cmd += "--cacert /etc/tls/certs/admin/tls.crt "
+		}
+	}
+	cmd += fmt.Sprintf("%s://${POD_NAME}.%s.%s.svc.cluster.local:%d", proto, r.pandaCluster.Name, r.pandaCluster.Namespace, adminAPI.Port)
+	cmd += "/v1/node_config"
+	return cmd
+}
+
+func (r *SecretResource) composeCURLMaintenanceCommand(
+	options string, urlOverwrite *string,
+) string {
+	adminAPI := r.pandaCluster.AdminAPIInternal()
+
+	cmd := fmt.Sprintf(`curl %s `, options)
+
+	tlsConfig := adminAPI.GetTLS()
+	proto := "http"
+	if tlsConfig != nil && tlsConfig.Enabled {
+		proto = "https"
+		if tlsConfig.RequireClientAuth {
+			cmd += "--cacert /etc/tls/certs/admin/ca/ca.crt --cert /etc/tls/certs/admin/tls.crt --key /etc/tls/certs/admin/tls.key "
```

Comment:
nit: would be good to de-duplicate these. Could be a follow-up

Reply:
totally agree